backend/src/ee/routes/v1/oidc-router.ts

```diff
@@ -0,0 +1,349 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-unsafe-return */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+// All the any rules are disabled because passport typesense with fastify is really poor
+
+import { Authenticator, Strategy } from "@fastify/passport";
+import fastifySession from "@fastify/session";
+import RedisStore from "connect-redis";
+import { Redis } from "ioredis";
+import { z } from "zod";
+
+import { OidcConfigsSchema } from "@app/db/schemas/oidc-configs";
+import { OIDCConfigurationType } from "@app/ee/services/oidc/oidc-config-types";
+import { getConfig } from "@app/lib/config/env";
+import { authRateLimit, readLimit, writeLimit } from "@app/server/config/rateLimiter";
+import { verifyAuth } from "@app/server/plugins/auth/verify-auth";
+import { AuthMode } from "@app/services/auth/auth-type";
+
+export const registerOidcRouter = async (server: FastifyZodProvider) => {
+  const appCfg = getConfig();
+  const redis = new Redis(appCfg.REDIS_URL);
+  const passport = new Authenticator({ key: "oidc", userProperty: "passportUser" });
+
+  /*
+  - OIDC protocol cannot work without sessions: https://github.com/panva/node-openid-client/issues/190
+  - Current redis usage is not ideal and will eventually have to be refactored to use a better structure
+  - Fastify session <> Redis structure is based on the ff: https://github.com/fastify/session/blob/master/examples/redis.js
+  */
+  const redisStore = new RedisStore({
+    client: redis,
+    prefix: "oidc-session:",
+    ttl: 600 // 10 minutes
+  });
+
+  await server.register(fastifySession, {
+    secret: appCfg.COOKIE_SECRET_SIGN_KEY,
+    store: redisStore,
+    cookie: {
+      secure: appCfg.HTTPS_ENABLED,
+      sameSite: "lax" // we want cookies to be sent to Infisical in redirects originating from IDP server
+    }
+  });
+
+  await server.register(passport.initialize());
+  await server.register(passport.secureSession());
+
+  // redirect to IDP for login
+  server.route({
+    url: "/login",
+    method: "GET",
+    config: {
+      rateLimit: authRateLimit
+    },
+    schema: {
+      querystring: z.object({
+        orgSlug: z.string().trim(),
+        callbackPort: z.string().trim().optional()
+      })
+    },
+    handler: async (req, res) => {
```

Comment:
Nit: Getting this error in the server logs here:

```
infisical-dev-api              |     err: {
infisical-dev-api              |       "type": "FastifyError",
infisical-dev-api              |       "message": "Reply was already sent, did you forget to \"return reply\" in \"/api/v1/sso/oidc/login?orgSlug=admin-org-xxx\" (GET)?",
infisical-dev-api              |       "stack":
infisical-dev-api              |           FastifyError: Reply was already sent, did you forget to "return reply" in "/api/v1/sso/oidc/login?orgSlug=admin-org-f-ta5" (GET)?
infisical-dev-api              |               at Reply.send (/app/node_modules/fastify/lib/reply.js:151:26)
infisical-dev-api              |               at Object.<anonymous> (/app/src/server/plugins/error-handler.ts:2:2530)
infisical-dev-api              |               at handleError (/app/node_modules/fastify/lib/error-handler.js:65:18)
infisical-dev-api              |               at onErrorHook (/app/node_modules/fastify/lib/reply.js:839:5)
infisical-dev-api              |               at wrapOnSendEnd (/app/node_modules/fastify/lib/reply.js:563:5)
infisical-dev-api              |               at next (/app/node_modules/fastify/lib/hooks.js:280:7)
infisical-dev-api              |               at /app/node_modules/@fastify/session/lib/fastifySession.js:193:11
infisical-dev-api              |               at /app/node_modules/@fastify/session/lib/session.js:176:11
infisical-dev-api              |               at RedisStore.set (/app/node_modules/connect-redis/dist/cjs/index.js:79:20)
infisical-dev-api              |               at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
infisical-dev-api              |       "code": "FST_ERR_REP_ALREADY_SENT",
infisical-dev-api              |       "name": "FastifyError",
infisical-dev-api              |       "statusCode": 500
infisical-dev-api              |     }
```

Doesn't seem to affect OIDC functionality but pollutes the server logs on every login.

Reply:
resolved ^